In [619]:
import pandas as pd
import numpy as np

# Response Variable

### Arrangement of To-be-used dataset for Response Variable
Training data
- 2011/12 ~ 2017/2018: NBASalaryData03-17.csv (457.57)
- 2018/19: contract_final (474)
- 2019/20: nba2k20-full.csv (429)
- 2020/21: contract_sql (523)
Testing data
- 2021/22: contract_basketball_reference.txt (476 rows)

### contract_basketball_reference.txt
Source: https://www.basketball-reference.com/contracts/players.html <br>
Year: 2021-22 and the following years if have a contract. <br>
Usage: Testing data

In [620]:
# pwd
contract_bb = pd.read_csv("../dataset/contract_basketball_reference.txt", header = 1, sep = ",")
contract_bb

Rk                       Player   Tm    2021-22    2022-23    2023-24  \
0      NaN                          NaN  NaN        NaN        NaN        NaN   
1      1.0      Stephen Curry\curryst01  GSW  $45780966  $48070014  $51915615   
2      2.0           John Wall\walljo01  HOU  $44310840  $47366760        NaN   
3      3.0  Russell Westbrook\westbru01  LAL  $44211146  $47063478        NaN   
4      4.0       James Harden\hardeja01  BRK  $43848000  $46872000        NaN   
..     ...                          ...  ...        ...        ...        ...   
471  471.0         Sam Dekker\dekkesa01  TOR    $350000        NaN        NaN   
472  472.0      Ish Wainright\wainris01  TOR        NaN        NaN        NaN   
473  473.0      Troy Williams\willitr02  HOU    $122741    $122741        NaN   
474  474.0      Jabari Parker\parkeja01  BOS   $2239544        NaN        NaN   
475  475.0  Demetrius Jackson\jacksde01  BOS     $92857     $92857     $92857   

       2024-25    2025-26 2026-27 Signed Using  Guaranteed  
0          NaN        NaN     NaN          NaN         NaN  
1    $55761216  $59606817     NaN  Bird Rights  $261134628  
2          NaN        NaN     NaN  Bird Rights   $44310840  
3          NaN        NaN     NaN  Bird Rights   $44211146  
4          NaN        NaN     NaN  Bird Rights   $43848000  
..         ...        ...     ...          ...         ...  
471        NaN        NaN     NaN          NaN     $350000  
472        NaN        NaN     NaN          NaN     $375000  
473        NaN        NaN     NaN          NaN     $245482  
474        NaN        NaN     NaN          NaN     $100000  
475        NaN        NaN     NaN          NaN     $278571  

[476 rows x 11 columns]

### nba_contracts_history.csv
Source: https://www.kaggle.com/jarosawjaworski/current-nba-players-contracts-history/version/1 <br>
Year: 2010/2011 season to 2019/2020 season.<br>
Usage: Non-used. <br>
Note:
- Only includes players that are active in 2020/2021 season.
- Doesn't include rookie or retained contracts.
- Doesn't include contracts for player that haven't played year before the signing the contract.


In [621]:
contract_kaggle = pd.read_csv("../dataset/nba_contracts_history.csv")
# contract_kaggle.AVG_SALARY.value_counts()
# contract_kaggle.AVG_SALARY.sort_values()
contract_kaggle
# contract_kaggle.head()

NAME  CONTRACT_START  CONTRACT_END    AVG_SALARY   AGE    GP  \
0    Wesley Matthews            2019          2020  2.564753e+06  32.0  69.0   
1        Brook Lopez            2015          2017  2.116568e+07  27.0  72.0   
2     DeAndre Jordan            2011          2014  1.075976e+07  22.0  80.0   
3    Markieff Morris            2015          2018  8.143324e+06  25.0  82.0   
4      Dwight Howard            2018          2019  1.341074e+07  32.0  81.0   
..               ...             ...           ...           ...   ...   ...   
194      Brook Lopez            2012          2014  1.469367e+07  24.0   5.0   
195   Nikola Vucevic            2015          2018  1.200000e+07  24.0  74.0   
196      Aron Baynes            2015          2017  5.766667e+06  28.0  70.0   
197   Andre Iguodala            2013          2016  1.200000e+07  29.0  80.0   
198   Draymond Green            2015          2019  1.640000e+07  25.0  79.0   

        W     L     MIN     PTS  ...   FT%   OREB   DREB     REB    AST  \
0    27.0  42.0  2091.0   840.0  ...  81.0   32.0  138.0   170.0  160.0   
1    34.0  38.0  2100.0  1236.0  ...  81.4  214.0  321.0   535.0   50.0   
2    31.0  49.0  2047.0   566.0  ...  45.2  207.0  368.0   575.0   42.0   
3    39.0  43.0  2581.0  1258.0  ...  76.3  110.0  396.0   506.0  192.0   
4    35.0  46.0  2463.0  1347.0  ...  57.4  255.0  757.0  1012.0  105.0   
..    ...   ...     ...     ...  ...   ...    ...    ...     ...    ...   
194   2.0   3.0   136.0    96.0  ...  62.5    8.0   10.0    18.0    6.0   
195  21.0  53.0  2529.0  1428.0  ...  75.2  238.0  572.0   810.0  147.0   
196  49.0  21.0  1122.0   461.0  ...  86.5  109.0  209.0   318.0   35.0   
197  57.0  23.0  2779.0  1038.0  ...  57.4   82.0  341.0   423.0  433.0   
198  64.0  15.0  2490.0   921.0  ...  66.0  114.0  533.0   647.0  291.0   

       TOV    STL    BLK     PF    +/-  
0     91.0   54.0   17.0  160.0  -90.0  
1    104.0   43.0  126.0  206.0 -104.0  
2    101.0   41.0  142.0  255.0 -100.0  
3    170.0  101.0   39.0  245.0  105.0  
4    208.0   48.0  131.0  249.0   89.0  
..     ...    ...    ...    ...    ...  
194    6.0    1.0    4.0    8.0  -27.0  
195  148.0   54.0   54.0  220.0 -302.0  
196   65.0   16.0   22.0  162.0  136.0  
197  206.0  139.0   52.0  127.0  380.0  
198  133.0  123.0   99.0  253.0  839.0  

[199 rows x 28 columns]

### NBASalaryData03-17.csv
Source: https://www.kaggle.com/hultm28/nba-player-salary-data-2002-2017 <br>
Year: 2002/2003 season to 2017/2018 season.<br>
Note: sourced from ESPN<br>
Usage: Training data-1

In [622]:
contract_kaggle_2 = pd.read_csv("../dataset/NBASalaryData03-17.csv")
contract_kaggle_2.loc[contract_kaggle_2.season == "2017-2018"]
#contract_kaggle_2

team      salary           player position     season
5792   Golden State Warriors  37457154.0    Stephen Curry       PG  2017-2018
5793          Boston Celtics  27734406.0       Al Horford        C  2017-2018
5794   Oklahoma City Thunder  26243760.0  Carmelo Anthony       PF  2017-2018
5795  Portland Trail Blazers  26153057.0   Damian Lillard       PG  2017-2018
5796      Washington Wizards  26011913.0  Otto Porter Jr.       SF  2017-2018
...                      ...         ...              ...      ...        ...
6250          Boston Celtics     44495.0  Jonathan Gibson        G  2017-2018
6251        Dallas Mavericks     31969.0   Pierre Jackson       PG  2017-2018
6252      Los Angeles Lakers     13824.0     Andre Ingram        G  2017-2018
6253           Orlando Magic      4608.0  Amile Jefferson       PF  2017-2018
6254         Houston Rockets      4608.0    Aaron Jackson        F  2017-2018

[463 rows x 5 columns]

In [623]:
print (len(contract_kaggle_2.loc[contract_kaggle_2.season == "2011-2012"]))
print (len(contract_kaggle_2.loc[contract_kaggle_2.season == "2012-2013"]))
print (len(contract_kaggle_2.loc[contract_kaggle_2.season == "2013-2014"]))
print (len(contract_kaggle_2.loc[contract_kaggle_2.season == "2014-2015"]))
print (len(contract_kaggle_2.loc[contract_kaggle_2.season == "2015-2016"]))
print (len(contract_kaggle_2.loc[contract_kaggle_2.season == "2016-2017"]))
print (len(contract_kaggle_2.loc[contract_kaggle_2.season == "2017-2018"]))
print ((494 + 531 + 385 + 435 + 433 + 462 + 463)/7)

494
531
385
435
433
462
463
457.57142857142856


### Player - Salaries per Year (1990 - 2017).xlsx
Source: https://www.kaggle.com/whitefero/nba-player-salary-19902017/version/1 <br>
Year: 1990/1991 season to 2017/2018 season.<br>
Note: sourced from ESPN<br>
Usage: Training data -3

In [624]:
contract_kaggle_3 = pd.read_excel("../dataset/Player - Salaries per Year (1990 - 2017).xlsx")
contract_kaggle_3.loc[contract_kaggle_3.iloc[:,3] == 2017]
# contract_kaggle_3

Register Value    Player Name  Salary in $  Season Start  Season End  \
20              11326   A.J. Hammons      1312611          2017        2018   
36              11327   Aaron Brooks      2116955          2017        2018   
41              11328   Aaron Gordon      5504420          2017        2018   
52              11329     Aaron Gray       452059          2017        2018   
83              11330    Abdel Nader      1167333          2017        2018   
...               ...            ...          ...           ...         ...   
11761           11833   Zach Collins      3057240          2017        2018   
11765           11834    Zach LaVine      3202217          2017        2018   
11782           11835  Zach Randolph     12307692          2017        2018   
11806           11836  Zaza Pachulia      3477600          2017        2018   
11817           11837        Zhou Qi       815615          2017        2018   

      Team          Full Team Name  
20     MIA              Miami Heat  
36     MIN  Minnesota Timberwolves  
41     ORL           Orlando Magic  
52     DET         Detroit Pistons  
83     BOS           Boston Celtic  
...    ...                     ...  
11761  POR  Portland Trail Blazers  
11765  CHI           Chicago Bulls  
11782  SAC        Sacramento Kings  
11806  GSW   Golden State Warriors  
11817  HOU         Houston Rockets  

[512 rows x 7 columns]

In [625]:
print (len(contract_kaggle_3.loc[contract_kaggle_3.iloc[:,3] == 2011]))
print (len(contract_kaggle_3.loc[contract_kaggle_3.iloc[:,3] == 2012]))
print (len(contract_kaggle_3.loc[contract_kaggle_3.iloc[:,3] == 2013]))
print (len(contract_kaggle_3.loc[contract_kaggle_3.iloc[:,3] == 2014]))
print (len(contract_kaggle_3.loc[contract_kaggle_3.iloc[:,3] == 2015]))
print (len(contract_kaggle_3.loc[contract_kaggle_3.iloc[:,3] == 2016]))
print (len(contract_kaggle_3.loc[contract_kaggle_3.iloc[:,3] == 2017]))
print ((453 + 429 + 397 + 527 + 563 + 609 + 512)/7)

453
429
397
527
563
609
512
498.57142857142856


### salaries_1985to2018.csv
Source: https://data.world/datadavis/nba-salaries <br>
Year: 1984/1985 season to 2017/2018 season.<br>
Usage: Training data-2

In [626]:
contract_dataworld_1985_2018 = pd.read_csv("../dataset/data.world/salaries_1985to2018.csv")
contract_dataworld_1985_2018.loc[contract_dataworld_1985_2018.season_start == 1984]
# contract_dataworld_1985_2018

league  player_id   salary   season  season_end  season_start  \
5        NBA  abdulka01  1530000  1984-85        1985          1984   
56       NBA  adamsal01   500000  1984-85        1985          1984   
104      NBA  aguirma01   800000  1984-85        1985          1984   
114      NBA  aingeda01   400000  1984-85        1985          1984   
660      NBA  baglejo01   209000  1984-85        1985          1984   
...      ...        ...      ...      ...         ...           ...   
13949    NBA   woodal01   450000  1984-85        1985          1984   
13960    NBA   woodle01   215000  1984-85        1985          1984   
13977    NBA  woodsmi01   375000  1984-85        1985          1984   
13982    NBA  woolror01   300000  1984-85        1985          1984   
14020    NBA  worthja01   400000  1984-85        1985          1984   

                      team  
5       Los Angeles Lakers  
56            Phoenix Suns  
104       Dallas Mavericks  
114         Boston Celtics  
660    Cleveland Cavaliers  
...                    ...  
13949  Seattle SuperSonics  
13960   Philadelphia 76ers  
13977    Kansas City Kings  
13982        Chicago Bulls  
14020   Los Angeles Lakers  

[210 rows x 7 columns]

### nba2k20-full.csv
Source: https://www.kaggle.com/isaienkov/nba2k20-player-dataset?select=nba2k20-full.csv <br>
Year: 2019/2020<br>
Usage: Training data

In [627]:
contract_2k20 = pd.read_csv("../dataset/nba2k20-full.csv")
contract_2k20

full_name  rating jersey                  team position  \
0             LeBron James      97    #23    Los Angeles Lakers        F   
1            Kawhi Leonard      97     #2  Los Angeles Clippers        F   
2    Giannis Antetokounmpo      96    #34       Milwaukee Bucks      F-G   
3             Kevin Durant      96     #7         Brooklyn Nets        F   
4             James Harden      96    #13       Houston Rockets        G   
..                     ...     ...    ...                   ...      ...   
424        Brandon Goodwin      68     #0         Atlanta Hawks        G   
425        Michael Frazier      68    #21       Houston Rockets        G   
426   Kostas Antetokounmpo      68    #37    Los Angeles Lakers        F   
427           Jalen Lecque      68     #0          Phoenix Suns        G   
428        Duncan Robinson      67    #55            Miami Heat        F   

        b_day       height                weight     salary country  \
0    12/30/84   6-9 / 2.06  250 lbs. / 113.4 kg.  $37436858     USA   
1    06/29/91   6-7 / 2.01  225 lbs. / 102.1 kg.  $32742000     USA   
2    12/06/94  6-11 / 2.11  242 lbs. / 109.8 kg.  $25842697  Greece   
3    09/29/88  6-10 / 2.08  230 lbs. / 104.3 kg.  $37199000     USA   
4    08/26/89   6-5 / 1.96   220 lbs. / 99.8 kg.  $38199000     USA   
..        ...          ...                   ...        ...     ...   
424  10/02/95   6-0 / 1.83   180 lbs. / 81.6 kg.    $522738     USA   
425  03/08/94   6-3 / 1.91   200 lbs. / 90.7 kg.     $79568     USA   
426  11/20/97  6-10 / 2.08   200 lbs. / 90.7 kg.     $79568  Greece   
427  06/13/00   6-3 / 1.91   190 lbs. / 86.2 kg.    $898310     USA   
428  04/22/94   6-7 / 2.01   215 lbs. / 97.5 kg.   $1416852     USA   

     draft_year draft_round draft_peak             college  
0          2003           1          1                 NaN  
1          2011           1         15     San Diego State  
2          2013           1         15                 NaN  
3          2007           1          2               Texas  
4          2009           1          3       Arizona State  
..          ...         ...        ...                 ...  
424        2018   Undrafted  Undrafted  Florida Gulf Coast  
425        2015           2         60             Florida  
426        2018   Undrafted  Undrafted              Dayton  
427        2019   Undrafted  Undrafted                 NaN  
428        2018   Undrafted  Undrafted            Michigan  

[429 rows x 14 columns]

### nba_final.csv
Source: https://www.kaggle.com/davra98/nba-players-20162019?select=nba_final.csv <br>
Year: 2016/2017 to 2018/19<br>
Usage: Training data

In [628]:
contract_final = pd.read_csv("../dataset/nba_final.csv")
contract_final.columns
#contract_final

Index(['Rk', 'Player.x', 'Player_ID', 'Pos1', 'Pos2', 'Age', 'Tm', 'G', 'GS',
       'MP', 'FG', 'FGA', 'FG.', 'X3P', 'X3PA', 'X3P.', 'X2P', 'X2PA', 'X2P.',
       'eFG.', 'FT', 'FTA', 'FT.', 'ORB', 'DRB', 'TRB', 'AST', 'STL', 'BLK',
       'TOV', 'PF', 'PTS', 'Salary', 'mean_views', 'Season', 'Conference',
       'Role', 'Fvot', 'FRank', 'Pvot', 'PRank', 'Mvot', 'MRank', 'Score',
       'Play'],
      dtype='object')

In [629]:
contract_final.loc[:,["Season", "Salary"]]
print (sum(contract_final.loc[:,"Season"] == "2016-17"))
print (sum(contract_final.loc[:,"Season"] == "2017-18"))
print (sum(contract_final.loc[:,"Season"] == "2018-19"))

436
498
474


### basketball.sqlite
Source: https://www.kaggle.com/wyattowalsh/basketball <br>
Year: 2020/2021 to the following years<br>
Usage: Training data

In [630]:
import sqlite3

# Read sqlite query results into a pandas DataFrame
con = sqlite3.connect("../dataset/1GB/archive/basketball.sqlite")
contract_sql = pd.read_sql_query("SELECT * from Player_Salary", con)
# print(contract_sql)

print (type(contract_sql))

con.close()

<class 'pandas.core.frame.DataFrame'>


In [631]:
contract_sql.slugSeason.value_counts()

2020-21    523
2021-22    363
2022-23    236
2023-24    121
2024-25     49
Name: slugSeason, dtype: int64

In [632]:
sum(contract_sql.slugSeason == "2020-21")
contract_sql.loc[contract_sql.slugSeason == "2020-21"]

slugSeason            nameTeam         namePlayer    statusPlayer  \
0       2020-21       Atlanta Hawks  Bogdan Bogdanovic  current roster   
4       2020-21       Atlanta Hawks    Brandon Goodwin  current roster   
6       2020-21       Atlanta Hawks     Bruno Fernando  current roster   
9       2020-21       Atlanta Hawks        Cam Reddish  current roster   
13      2020-21       Atlanta Hawks       Clint Capela  current roster   
...         ...                 ...                ...             ...   
1281    2020-21  Washington Wizards          Raul Neto  current roster   
1282    2020-21  Washington Wizards        Robin Lopez  current roster   
1283    2020-21  Washington Wizards      Rui Hachimura  current roster   
1287    2020-21  Washington Wizards  Russell Westbrook  current roster   
1290    2020-21  Washington Wizards      Thomas Bryant  current roster   

      isFinalSeason  isWaived  isOnRoster  isNonGuaranteed  isTeamOption  \
0                 0         0           1                0             0   
4                 0         0           1                0             0   
6                 0         0           1                0             0   
9                 0         0           1                0             0   
13                0         0           1                0             0   
...             ...       ...         ...              ...           ...   
1281              1         0           1                0             0   
1282              1         0           1                0             0   
1283              0         0           1                0             0   
1287              0         0           1                0             0   
1290              0         0           1                0             0   

      isPlayerOption typeContractDetail       value  
0                  0         Guaranteed  18000000.0  
4                  0         Guaranteed   1701593.0  
6                  0         Guaranteed   1517981.0  
9                  0         Guaranteed   4458000.0  
13                 0         Guaranteed  16000000.0  
...              ...                ...         ...  
1281               0         Guaranteed   1620564.0  
1282               0         Guaranteed   7300000.0  
1283               0         Guaranteed   4692840.0  
1287               0         Guaranteed  41358814.0  
1290               0         Guaranteed   8333333.0  

[523 rows x 12 columns]

### Arrangement of To-be-used dataset for Response Variable
Training data
- 2011/12 ~ 2017/2018: NBASalaryData03-17.csv (457.57) - contract_kaggle_2
- 2018/19: contract_final (474) - contract_final
- 2019/20: nba2k20-full.csv (429) - contract_2k20
- 2020/21: contract_sql (523) - contract_sql
Testing data
- 2021/22: contract_basketball_reference.txt (476 rows) - contract_bb

In [633]:
# handling 2012~2017 data
trim_contract_kaggle_2 = contract_kaggle_2.loc[:,['player', 'season', "salary"]]
trim_contract_kaggle_2.loc[:,'season'] = trim_contract_kaggle_2.loc[:,'season'].str.split('-', expand=True)[0]
trim_contract_kaggle_2['season'] = pd.to_numeric(trim_contract_kaggle_2['season'])
trim_contract_kaggle_2 = trim_contract_kaggle_2.loc[trim_contract_kaggle_2['season'] >= 2012]
trim_contract_kaggle_2

player  season      salary
3546      Kobe Bryant    2012  30453805.0
3547       Chris Paul    2012  18668431.0
3548     Kevin Durant    2012  17832626.0
3549     Derrick Rose    2012  17632688.0
3550   Deron Williams    2012  17177795.0
...               ...     ...         ...
6250  Jonathan Gibson    2017     44495.0
6251   Pierre Jackson    2017     31969.0
6252     Andre Ingram    2017     13824.0
6253  Amile Jefferson    2017      4608.0
6254    Aaron Jackson    2017      4608.0

[2709 rows x 3 columns]

In [634]:
# handling 2018 data
trim_contract_final = contract_final.loc[contract_final.loc[:, 'Season'] == "2018-19", ['Player.x', 'Season', 'Salary']]
trim_contract_final.rename(columns={'Player.x': 'player', 'Season': 'season', 'Salary': 'salary'}, inplace=True)
trim_contract_final.loc[:,'season'] = trim_contract_final.loc[:,'season'].str.split('-', expand=True)[0]
trim_contract_final['season'] = pd.to_numeric(trim_contract_final['season'])
trim_contract_final

player  season      salary
934      Aaron Gordon    2018  21590909.0
935     Aaron Holiday    2018   1914480.0
936       Abdel Nader    2018   1378242.0
937   Al-Farouq Aminu    2018   6957105.0
938        Al Horford    2018  28928710.0
...               ...     ...         ...
1403     Zach Collins    2018   3628920.0
1404      Zach LaVine    2018  19500000.0
1405      Zach Lofton    2018     77250.0
1406    Zaza Pachulia    2018   2393887.0
1407     Zhaire Smith    2018   2615160.0

[474 rows x 3 columns]

In [635]:
# handling 2019 data
trim_contract_2k20 = contract_2k20.loc[:,["full_name", "salary"]]
trim_contract_2k20.loc[:,'season'] = '2019'
trim_contract_2k20 = trim_contract_2k20[['full_name', 'season', 'salary']]
trim_contract_2k20.rename(columns={'full_name': 'player'}, inplace=True)
trim_contract_2k20['season'] = pd.to_numeric(trim_contract_2k20['season'])
trim_contract_2k20


player  season     salary
0             LeBron James    2019  $37436858
1            Kawhi Leonard    2019  $32742000
2    Giannis Antetokounmpo    2019  $25842697
3             Kevin Durant    2019  $37199000
4             James Harden    2019  $38199000
..                     ...     ...        ...
424        Brandon Goodwin    2019    $522738
425        Michael Frazier    2019     $79568
426   Kostas Antetokounmpo    2019     $79568
427           Jalen Lecque    2019    $898310
428        Duncan Robinson    2019   $1416852

[429 rows x 3 columns]

In [636]:
# handling 2020 data
trim_contract_sql = contract_sql.loc[:,['slugSeason', 'namePlayer', 'value']]
trim_contract_sql = trim_contract_sql.loc[trim_contract_sql['slugSeason'] == '2020-21']
trim_contract_sql.rename(columns={'namePlayer': 'player', 'slugSeason': 'season', 'value':'salary'}, inplace=True)
trim_contract_sql = trim_contract_sql[['player', 'season', 'salary']]
trim_contract_sql.loc[:,'season'] = trim_contract_sql.loc[:,'season'].str.split('-', expand=True)[0]
trim_contract_sql['season'] = pd.to_numeric(trim_contract_sql['season'])
trim_contract_sql


player  season      salary
0     Bogdan Bogdanovic    2020  18000000.0
4       Brandon Goodwin    2020   1701593.0
6        Bruno Fernando    2020   1517981.0
9           Cam Reddish    2020   4458000.0
13         Clint Capela    2020  16000000.0
...                 ...     ...         ...
1281          Raul Neto    2020   1620564.0
1282        Robin Lopez    2020   7300000.0
1283      Rui Hachimura    2020   4692840.0
1287  Russell Westbrook    2020  41358814.0
1290      Thomas Bryant    2020   8333333.0

[523 rows x 3 columns]

In [637]:
# handling 2021 data
trim_contract_bb = contract_bb.loc[:,['Player', '2021-22']]
trim_contract_bb = trim_contract_bb.iloc[1:, :]
trim_contract_bb.loc[:, 'season'] = '2021'
trim_contract_bb = trim_contract_bb[['Player', 'season', '2021-22']]
trim_contract_bb.rename(columns={'Player': 'player', 'season': 'season', '2021-22': 'salary'}, inplace=True)
trim_contract_bb.loc[:,'player'] = trim_contract_bb.loc[:,'player'].str.split('\\', expand=True)[0]
trim_contract_bb['season'] = pd.to_numeric(trim_contract_bb['season'])
trim_contract_bb

player  season     salary
1        Stephen Curry    2021  $45780966
2            John Wall    2021  $44310840
3    Russell Westbrook    2021  $44211146
4         James Harden    2021  $43848000
5       Damian Lillard    2021  $43750000
..                 ...     ...        ...
471         Sam Dekker    2021    $350000
472      Ish Wainright    2021        NaN
473      Troy Williams    2021    $122741
474      Jabari Parker    2021   $2239544
475  Demetrius Jackson    2021     $92857

[475 rows x 3 columns]

In [638]:
trim_contract_kaggle_2
# trim_contract_final

player  season      salary
3546      Kobe Bryant    2012  30453805.0
3547       Chris Paul    2012  18668431.0
3548     Kevin Durant    2012  17832626.0
3549     Derrick Rose    2012  17632688.0
3550   Deron Williams    2012  17177795.0
...               ...     ...         ...
6250  Jonathan Gibson    2017     44495.0
6251   Pierre Jackson    2017     31969.0
6252     Andre Ingram    2017     13824.0
6253  Amile Jefferson    2017      4608.0
6254    Aaron Jackson    2017      4608.0

[2709 rows x 3 columns]

In [639]:
df_salary = trim_contract_kaggle_2.append(trim_contract_final, sort=False)
df_salary = df_salary.append(trim_contract_2k20, sort=False)
df_salary = df_salary.append(trim_contract_sql, sort=False)
df_salary = df_salary.append(trim_contract_bb, sort=False)


df_salary

player  season      salary
3546        Kobe Bryant    2012  30453805.0
3547         Chris Paul    2012  18668431.0
3548       Kevin Durant    2012  17832626.0
3549       Derrick Rose    2012  17632688.0
3550     Deron Williams    2012  17177795.0
...                 ...     ...         ...
471          Sam Dekker    2021     $350000
472       Ish Wainright    2021         NaN
473       Troy Williams    2021     $122741
474       Jabari Parker    2021    $2239544
475   Demetrius Jackson    2021      $92857

[4610 rows x 3 columns]

In [640]:
# df_salary
# df_salary.season.value_counts()
# df_salary.isnull().sum()
# df_salary.loc[df_salary.salary.isnull()]

# Predictors

Method 1
- merge df_salary with predictor_1 (more year, less statistics) 
    - 2707 null data (root cause clarified; time consuming, better way?)
- merge df_salay with predictor_2 (less year, ~ 2016, more statistics) 
    - 16XX null data to be clarified >>> give up
    
Method 2 (preferred, suggested)
- separate predictor_1 into two halfs (~ 2016, 2017~)
    - merge predictor_1_2016 & predictor_2
        - only 941 rows left after inner
        - more if outter, but how to do
- concate the previous with predictor_1_2017

- many missing data >>> what to do?

In [641]:
all_kaggle1 = pd.read_csv("../dataset/age/all_seasons.csv")
all_kaggle1.columns


Index(['Unnamed: 0', 'player_name', 'team_abbreviation', 'age',
       'player_height', 'player_weight', 'college', 'country', 'draft_year',
       'draft_round', 'draft_number', 'gp', 'pts', 'reb', 'ast', 'net_rating',
       'oreb_pct', 'dreb_pct', 'usg_pct', 'ts_pct', 'ast_pct', 'season'],
      dtype='object')

In [642]:
all_kaggle1 = age_kaggle1.loc[:,['player_name', 'team_abbreviation','age', 'gp', 'pts', 'reb', 'ast','net_rating', 'season']]
all_kaggle1.loc[:,'season'] = all_kaggle1.loc[:,'season'].str.split('-', expand=True)[0]
all_kaggle1['season'] = pd.to_numeric(all_kaggle1['season'])
all_kaggle1.rename(columns={'player_name':'player','team_abbreviation': 'team', 'gp': 'games', 'pts': 'PTS', 'reb':'REB', 'ast':'AST', 'net_rating':'NET_RATING'}, inplace=True)
all_kaggle1 = all_kaggle1.loc[all_kaggle1['season'] >= 2012]



all_kaggle1

player team   age  games   PTS   REB  AST  NET_RATING  \
7156           Nolan Smith  POR  24.0     40   2.8   0.7  0.9        -4.2   
7157        Nikola Vucevic  ORL  22.0     77  13.1  11.9  1.9        -8.1   
7158           Norris Cole  MIA  24.0     80   5.6   1.6  2.1         3.1   
7159             Ray Allen  MIA  37.0     79  10.9   2.7  1.7         5.5   
7160             Omer Asik  HOU  26.0     82  10.1  11.7  0.9         5.9   
...                    ...  ...   ...    ...   ...   ...  ...         ...   
11695  Matthew Dellavedova  CLE  30.0     13   2.8   1.8  4.5        -3.1   
11696     Maurice Harkless  SAC  28.0     37   5.2   2.4  1.2        -2.9   
11697            Max Strus  MIA  25.0     39   6.1   1.1  0.6        -4.2   
11698    Marcus Morris Sr.  LAC  31.0     57  13.4   4.1  1.0         4.2   
11699         Aaron Gordon  DEN  25.0     50  12.4   5.7  3.2         2.1   

       season  
7156     2012  
7157     2012  
7158     2012  
7159     2012  
7160     2012  
...       ...  
11695    2020  
11696    2020  
11697    2020  
11698    2020  
11699    2020  

[4544 rows x 9 columns]

In [643]:
all_kaggle1.isnull().sum()

player        0
team          0
age           0
games         0
PTS           0
REB           0
AST           0
NET_RATING    0
season        0
dtype: int64

In [644]:
all_kaggle_2_players = pd.read_csv("../dataset/age/archive/Players.csv")
all_kaggle_2_players

Unnamed: 0             Player  height  weight  \
0              0    Curly Armstrong   180.0    77.0   
1              1       Cliff Barker   188.0    83.0   
2              2      Leo Barnhorst   193.0    86.0   
3              3         Ed Bartels   196.0    88.0   
4              4        Ralph Beard   178.0    79.0   
...          ...                ...     ...     ...   
3917        3917      Troy Williams   198.0    97.0   
3918        3918       Kyle Wiltjer   208.0   108.0   
3919        3919  Stephen Zimmerman   213.0   108.0   
3920        3920        Paul Zipser   203.0    97.0   
3921        3921        Ivica Zubac   216.0   120.0   

                              collage    born      birth_city  \
0                  Indiana University  1918.0             NaN   
1              University of Kentucky  1921.0        Yorktown   
2            University of Notre Dame  1924.0             NaN   
3     North Carolina State University  1925.0             NaN   
4              University of Kentucky  1927.0     Hardinsburg   
...                               ...     ...             ...   
3917  South Carolina State University  1969.0        Columbia   
3918               Gonzaga University  1992.0        Portland   
3919  University of Nevada, Las Vegas  1996.0  Hendersonville   
3920                              NaN  1994.0      Heidelberg   
3921                              NaN  1997.0          Mostar   

                 birth_state  
0                        NaN  
1                    Indiana  
2                        NaN  
3                        NaN  
4                   Kentucky  
...                      ...  
3917          South Carolina  
3918                  Oregon  
3919               Tennessee  
3920                 Germany  
3921  Bosnia and Herzegovina  

[3922 rows x 8 columns]

In [645]:
all_kaggle_2_season_stat = pd.read_csv("../dataset/age/archive/Seasons_Stats.csv")
all_kaggle_2_season_stat

Unnamed: 0    Year             Player  Pos   Age   Tm     G    GS  \
0               0  1950.0    Curly Armstrong  G-F  31.0  FTW  63.0   NaN   
1               1  1950.0       Cliff Barker   SG  29.0  INO  49.0   NaN   
2               2  1950.0      Leo Barnhorst   SF  25.0  CHS  67.0   NaN   
3               3  1950.0         Ed Bartels    F  24.0  TOT  15.0   NaN   
4               4  1950.0         Ed Bartels    F  24.0  DNN  13.0   NaN   
...           ...     ...                ...  ...   ...  ...   ...   ...   
24686       24686  2017.0        Cody Zeller   PF  24.0  CHO  62.0  58.0   
24687       24687  2017.0       Tyler Zeller    C  27.0  BOS  51.0   5.0   
24688       24688  2017.0  Stephen Zimmerman    C  20.0  ORL  19.0   0.0   
24689       24689  2017.0        Paul Zipser   SF  22.0  CHI  44.0  18.0   
24690       24690  2017.0        Ivica Zubac    C  19.0  LAL  38.0  11.0   

           MP   PER  ...    FT%    ORB    DRB    TRB    AST   STL   BLK   TOV  \
0         NaN   NaN  ...  0.705    NaN    NaN    NaN  176.0   NaN   NaN   NaN   
1         NaN   NaN  ...  0.708    NaN    NaN    NaN  109.0   NaN   NaN   NaN   
2         NaN   NaN  ...  0.698    NaN    NaN    NaN  140.0   NaN   NaN   NaN   
3         NaN   NaN  ...  0.559    NaN    NaN    NaN   20.0   NaN   NaN   NaN   
4         NaN   NaN  ...  0.548    NaN    NaN    NaN   20.0   NaN   NaN   NaN   
...       ...   ...  ...    ...    ...    ...    ...    ...   ...   ...   ...   
24686  1725.0  16.7  ...  0.679  135.0  270.0  405.0   99.0  62.0  58.0  65.0   
24687   525.0  13.0  ...  0.564   43.0   81.0  124.0   42.0   7.0  21.0  20.0   
24688   108.0   7.3  ...  0.600   11.0   24.0   35.0    4.0   2.0   5.0   3.0   
24689   843.0   6.9  ...  0.775   15.0  110.0  125.0   36.0  15.0  16.0  40.0   
24690   609.0  17.0  ...  0.653   41.0  118.0  159.0   30.0  14.0  33.0  30.0   

          PF    PTS  
0      217.0  458.0  
1       99.0  279.0  
2      192.0  438.0  
3       29.0   63.0  
4       27.0   59.0  
...      ...    ...  
24686  189.0  639.0  
24687   61.0  178.0  
24688   17.0   23.0  
24689   78.0  240.0  
24690   66.0  284.0  

[24691 rows x 53 columns]

In [646]:
all_kaggle_2_player_data = pd.read_csv("../dataset/age/archive/player_data.csv")
all_kaggle_2_player_data

name  year_start  year_end position height  weight  \
0          Alaa Abdelnaby        1991      1995      F-C   6-10   240.0   
1         Zaid Abdul-Aziz        1969      1978      C-F    6-9   235.0   
2     Kareem Abdul-Jabbar        1970      1989        C    7-2   225.0   
3      Mahmoud Abdul-Rauf        1991      2001        G    6-1   162.0   
4       Tariq Abdul-Wahad        1998      2003        F    6-6   223.0   
...                   ...         ...       ...      ...    ...     ...   
4545           Ante Zizic        2018      2018      F-C   6-11   250.0   
4546             Jim Zoet        1983      1983        C    7-1   240.0   
4547            Bill Zopf        1971      1971        G    6-1   170.0   
4548          Ivica Zubac        2017      2018        C    7-1   265.0   
4549           Matt Zunic        1949      1949      G-F    6-3   195.0   

             birth_date                                college  
0         June 24, 1968                        Duke University  
1         April 7, 1946                  Iowa State University  
2        April 16, 1947  University of California, Los Angeles  
3         March 9, 1969             Louisiana State University  
4      November 3, 1974              San Jose State University  
...                 ...                                    ...  
4545    January 4, 1997                                    NaN  
4546  December 20, 1953                  Kent State University  
4547       June 7, 1948                    Duquesne University  
4548     March 18, 1997                                    NaN  
4549  December 19, 1919           George Washington University  

[4550 rows x 8 columns]

In [647]:
all_kaggle_2_season_stat.columns
all_kaggle_2_extract = all_kaggle_2_season_stat.loc[:,['Year', 'Player', 'Age', 'Tm', 'G', 'PER', 'WS', 'BPM', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PTS']]
all_kaggle_2_extract.rename(columns={'Year': 'season', 'Player': 'player'}, inplace=True)
all_kaggle_2_extract['season'] = all_kaggle_2_extract['season'] - 1
all_kaggle_2_extract = all_kaggle_2_extract.loc[all_kaggle_2_extract['season'] >= 2012]

all_kaggle_2_extract

season             player   Age   Tm     G   PER   WS  BPM    TRB  \
21679  2012.0         Quincy Acy  22.0  TOR  29.0  15.9  1.1  0.7   77.0   
21680  2012.0        Jeff Adrien  26.0  CHA  52.0  13.4  1.0 -3.3  196.0   
21681  2012.0      Arron Afflalo  27.0  ORL  64.0  13.0  2.0 -2.3  239.0   
21682  2012.0       Josh Akognon  26.0  DAL   3.0  15.3  0.0 -0.2    1.0   
21683  2012.0       Cole Aldrich  24.0  TOT  45.0  11.1  0.6 -4.0  120.0   
...       ...                ...   ...  ...   ...   ...  ...  ...    ...   
24686  2016.0        Cody Zeller  24.0  CHO  62.0  16.7  5.6  2.1  405.0   
24687  2016.0       Tyler Zeller  27.0  BOS  51.0  13.0  1.0 -2.5  124.0   
24688  2016.0  Stephen Zimmerman  20.0  ORL  19.0   7.3  0.0 -7.3   35.0   
24689  2016.0        Paul Zipser  22.0  CHI  44.0   6.9  0.5 -3.7  125.0   
24690  2016.0        Ivica Zubac  19.0  LAL  38.0  17.0  1.1 -2.5  159.0   

         AST   STL   BLK    TOV     PTS  
21679   11.0  13.0  15.0   17.0   116.0  
21680   36.0  18.0  27.0   32.0   209.0  
21681  206.0  40.0  11.0  138.0  1057.0  
21682    1.0   0.0   0.0    0.0     5.0  
21683    9.0   5.0  23.0   23.0   100.0  
...      ...   ...   ...    ...     ...  
24686   99.0  62.0  58.0   65.0   639.0  
24687   42.0   7.0  21.0   20.0   178.0  
24688    4.0   2.0   5.0    3.0    23.0  
24689   36.0  15.0  16.0   40.0   240.0  
24690   30.0  14.0  33.0   30.0   284.0  

[3008 rows x 14 columns]

In [648]:
all_kaggle_2_extract.loc[all_kaggle_2_extract['player'] == 'Derrick Rose']

season        player   Age   Tm     G   PER   WS  BPM    TRB    AST  \
22728  2013.0  Derrick Rose  25.0  CHI  10.0   9.7 -0.2 -5.1   32.0   43.0   
23373  2014.0  Derrick Rose  26.0  CHI  51.0  15.9  1.2 -1.3  161.0  251.0   
23973  2015.0  Derrick Rose  27.0  CHI  66.0  13.5  0.4 -3.3  225.0  311.0   
24562  2016.0  Derrick Rose  28.0  NYK  64.0  17.0  3.0 -1.9  246.0  283.0   

        STL   BLK    TOV     PTS  
22728   5.0   1.0   34.0   159.0  
23373  36.0  16.0  161.0   904.0  
23973  44.0  14.0  175.0  1080.0  
24562  44.0  17.0  147.0  1154.0

In [649]:
all_kaggle_2_extract.isnull().sum()

season    0
player    0
Age       0
Tm        0
G         0
PER       1
WS        0
BPM       0
TRB       0
AST       0
STL       0
BLK       0
TOV       0
PTS       0
dtype: int64

In [650]:
# df_temp_salary_kaggle2
df_temp_salary_kaggle2 = pd.merge(df_salary, all_kaggle_2_extract, on=['player','season'], how = 'left')
# df_all_data.loc[df_all_data['player'] == 'Kobe Bryant']
df_temp_salary_kaggle2.rename(columns={'Age': 'age', 'Tm': 'team', 'G': 'games', 'TRB': 'REB'}, inplace=True)
df_temp_salary_kaggle2 = df_temp_salary_kaggle2[['player', 'season', 'salary', 'team', 'games', 'age', 'PER', 'WS', 'BPM','PTS', 'REB', 'AST', 'STL', 'BLK', 'TOV']]
df_temp_salary_kaggle2['PTS'] = round(df_temp_salary_kaggle2['PTS'] / df_temp_salary_kaggle2['games'], 1)
df_temp_salary_kaggle2['REB'] = round(df_temp_salary_kaggle2['REB'] / df_temp_salary_kaggle2['games'], 1)
df_temp_salary_kaggle2['AST'] = round(df_temp_salary_kaggle2['AST'] / df_temp_salary_kaggle2['games'], 1)
df_temp_salary_kaggle2['STL'] = round(df_temp_salary_kaggle2['STL'] / df_temp_salary_kaggle2['games'], 1)
df_temp_salary_kaggle2['BLK'] = round(df_temp_salary_kaggle2['BLK'] / df_temp_salary_kaggle2['games'], 1)
df_temp_salary_kaggle2['TOV'] = round(df_temp_salary_kaggle2['TOV'] / df_temp_salary_kaggle2['games'], 1)

df_temp_salary_kaggle2

player  season      salary team  games   age   PER    WS  \
0           Kobe Bryant    2012  30453805.0  LAL   78.0  34.0  23.0  10.9   
1            Chris Paul    2012  18668431.0  LAC   70.0  27.0  26.4  13.9   
2          Kevin Durant    2012  17832626.0  OKC   81.0  24.0  28.3  18.9   
3          Derrick Rose    2012  17632688.0  NaN    NaN   NaN   NaN   NaN   
4        Deron Williams    2012  17177795.0  BRK   78.0  28.0  20.3  10.9   
...                 ...     ...         ...  ...    ...   ...   ...   ...   
5096         Sam Dekker    2021     $350000  NaN    NaN   NaN   NaN   NaN   
5097      Ish Wainright    2021         NaN  NaN    NaN   NaN   NaN   NaN   
5098      Troy Williams    2021     $122741  NaN    NaN   NaN   NaN   NaN   
5099      Jabari Parker    2021    $2239544  NaN    NaN   NaN   NaN   NaN   
5100  Demetrius Jackson    2021      $92857  NaN    NaN   NaN   NaN   NaN   

      BPM   PTS  REB  AST  STL  BLK  TOV  
0     4.7  27.3  5.6  6.0  1.4  0.3  3.7  
1     7.5  16.9  3.7  9.7  2.4  0.1  2.3  
2     7.7  28.1  7.9  4.6  1.4  1.3  3.5  
3     NaN   NaN  NaN  NaN  NaN  NaN  NaN  
4     3.5  18.9  3.0  7.7  1.0  0.4  2.8  
...   ...   ...  ...  ...  ...  ...  ...  
5096  NaN   NaN  NaN  NaN  NaN  NaN  NaN  
5097  NaN   NaN  NaN  NaN  NaN  NaN  NaN  
5098  NaN   NaN  NaN  NaN  NaN  NaN  NaN  
5099  NaN   NaN  NaN  NaN  NaN  NaN  NaN  
5100  NaN   NaN  NaN  NaN  NaN  NaN  NaN  

[5101 rows x 15 columns]

In [651]:
#try outter
temp = pd.merge(df_salary, all_kaggle_2_extract, on=['player','season'], how = 'inner')
# df_all_data.loc[df_all_data['player'] == 'Kobe Bryant']
temp.rename(columns={'Age': 'age', 'Tm': 'team', 'G': 'games', 'TRB': 'REB'}, inplace=True)
temp = df_temp_salary_kaggle2[['player', 'season', 'salary', 'team', 'games', 'age', 'PER', 'WS', 'BPM','PTS', 'REB', 'AST', 'STL', 'BLK', 'TOV']]
temp['PTS'] = round(temp['PTS'] / temp['games'], 1)
temp['REB'] = round(temp['REB'] / temp['games'], 1)
temp['AST'] = round(temp['AST'] / temp['games'], 1)
temp['STL'] = round(temp['STL'] / temp['games'], 1)
temp['BLK'] = round(temp['BLK'] / temp['games'], 1)
temp['TOV'] = round(temp['TOV'] / temp['games'], 1)

temp

player  season      salary team  games   age   PER    WS  \
0           Kobe Bryant    2012  30453805.0  LAL   78.0  34.0  23.0  10.9   
1            Chris Paul    2012  18668431.0  LAC   70.0  27.0  26.4  13.9   
2          Kevin Durant    2012  17832626.0  OKC   81.0  24.0  28.3  18.9   
3          Derrick Rose    2012  17632688.0  NaN    NaN   NaN   NaN   NaN   
4        Deron Williams    2012  17177795.0  BRK   78.0  28.0  20.3  10.9   
...                 ...     ...         ...  ...    ...   ...   ...   ...   
5096         Sam Dekker    2021     $350000  NaN    NaN   NaN   NaN   NaN   
5097      Ish Wainright    2021         NaN  NaN    NaN   NaN   NaN   NaN   
5098      Troy Williams    2021     $122741  NaN    NaN   NaN   NaN   NaN   
5099      Jabari Parker    2021    $2239544  NaN    NaN   NaN   NaN   NaN   
5100  Demetrius Jackson    2021      $92857  NaN    NaN   NaN   NaN   NaN   

      BPM  PTS  REB  AST  STL  BLK  TOV  
0     4.7  0.4  0.1  0.1  0.0  0.0  0.0  
1     7.5  0.2  0.1  0.1  0.0  0.0  0.0  
2     7.7  0.3  0.1  0.1  0.0  0.0  0.0  
3     NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4     3.5  0.2  0.0  0.1  0.0  0.0  0.0  
...   ...  ...  ...  ...  ...  ...  ...  
5096  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
5097  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
5098  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
5099  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
5100  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[5101 rows x 15 columns]

In [652]:
temp.isnull().sum()

player       0
season       0
salary      25
team      2706
games     2706
age       2706
PER       2707
WS        2706
BPM       2706
PTS       2706
REB       2706
AST       2706
STL       2706
BLK       2706
TOV       2706
dtype: int64

### Predictor first investigation on the null 2707 data
- df_salary
    - from 2012 ~ 2021
    - 4610 rows
    - 25 null data in salary

- all_kaggle_2_extract
    - from 2012 ~ 2016
    - 'Age', 'Tm', 'G','PTS', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PER', 'WS', 'BPM',
    - 3008 rows
    - no null data
- combine method
    - df_temp_salary_kaggle2 = pd.merge(df_salary, all_kaggle_2_extract, on=['player','season'], how = 'left')


#### Conclusion
- 2012 ~ 2016: 343 observations missed due to data incompleteness in either df_salary or all_kaggle_2_extract
- 2017 ~ : 2364 observations missed due to NO data in all_kaggle_2_extract
- Result: Reasonable


In [653]:
# df_temp_salary_kaggle2.isnull().sum()
# df_temp_salary_kaggle2.loc[]
# df_temp_salary_kaggle2.salary.isnull().sum()
# df_temp_salary_kaggle2[df_temp_salary_kaggle2[["player", 'season', 'salary']].isnull()]

print (df_temp_salary_kaggle2.loc[df_temp_salary_kaggle2.isnull().any(axis=1)].loc[df_temp_salary_kaggle2.season <= 2016])
df_temp_salary_kaggle2.loc[df_temp_salary_kaggle2.isnull().any(axis=1)].loc[df_temp_salary_kaggle2.season <= 2016].head(10)



                   player  season      salary team  games  age  PER  WS  BPM  \
3            Derrick Rose    2012  17632688.0  NaN    NaN  NaN  NaN NaN  NaN   
5            Andrew Bynum    2012  16889000.0  NaN    NaN  NaN  NaN NaN  NaN   
52            Mehmet Okur    2012  10890000.0  NaN    NaN  NaN  NaN NaN  NaN   
62         Cuttino Mobley    2012   9500000.0  NaN    NaN  NaN  NaN NaN  NaN   
84         Andres Nocioni    2012   6650000.0  NaN    NaN  NaN  NaN NaN  NaN   
...                   ...     ...         ...  ...    ...  ...  ...  ..  ...   
2657      Jonathan Holmes    2016    543471.0  NaN    NaN  NaN  NaN NaN  NaN   
2658          Sheldon Mac    2016    543471.0  NaN    NaN  NaN  NaN NaN  NaN   
2685  Johnny O'Bryant III    2016    161483.0  NaN    NaN  NaN  NaN NaN  NaN   
2696         Ray McCallum    2016    121112.0  NaN    NaN  NaN  NaN NaN  NaN   
2725       Gary Payton II    2016     35116.0  NaN    NaN  NaN  NaN NaN  NaN   

      PTS  REB  AST  STL  BLK  TOV  
3 

player  season      salary team  games  age  PER  WS  BPM  \
3         Derrick Rose    2012  17632688.0  NaN    NaN  NaN  NaN NaN  NaN   
5         Andrew Bynum    2012  16889000.0  NaN    NaN  NaN  NaN NaN  NaN   
52         Mehmet Okur    2012  10890000.0  NaN    NaN  NaN  NaN NaN  NaN   
62      Cuttino Mobley    2012   9500000.0  NaN    NaN  NaN  NaN NaN  NaN   
84      Andres Nocioni    2012   6650000.0  NaN    NaN  NaN  NaN NaN  NaN   
85        Jerome James    2012   6600000.0  NaN    NaN  NaN  NaN NaN  NaN   
89       Channing Frye    2012   6400000.0  NaN    NaN  NaN  NaN NaN  NaN   
105      Sasha Vujacic    2012   5475113.0  NaN    NaN  NaN  NaN NaN  NaN   
107        Mark Blount    2012   7967375.0  NaN    NaN  NaN  NaN NaN  NaN   
110  Metta World Peace    2012   7727500.0  NaN    NaN  NaN  NaN NaN  NaN   

     PTS  REB  AST  STL  BLK  TOV  
3    NaN  NaN  NaN  NaN  NaN  NaN  
5    NaN  NaN  NaN  NaN  NaN  NaN  
52   NaN  NaN  NaN  NaN  NaN  NaN  
62   NaN  NaN  NaN  NaN  NaN  NaN  
84   NaN  NaN  NaN  NaN  NaN  NaN  
85   NaN  NaN  NaN  NaN  NaN  NaN  
89   NaN  NaN  NaN  NaN  NaN  NaN  
105  NaN  NaN  NaN  NaN  NaN  NaN  
107  NaN  NaN  NaN  NaN  NaN  NaN  
110  NaN  NaN  NaN  NaN  NaN  NaN

In [654]:
print (df_salary.loc[(df_salary.loc[:, 'player'] == 'Derrick Rose')])
print ()
print (df_salary.loc[(df_salary.loc[:, 'player'] == 'Derrick Rose') & (df_salary.loc[:, 'season'] == 2012)])
all_kaggle_2_extract.loc[(all_kaggle_2_extract.loc[:, 'player'] == 'Derrick Rose')]

# print (df_salary.loc[(df_salary.loc[:, 'player'] == 'Sasha Vujacic')])
# print ()
# print (df_salary.loc[(df_salary.loc[:, 'player'] == 'Sasha Vujacic') & (df_salary.loc[:, 'season'] == 2016)])
# all_kaggle_2_extract.loc[(all_kaggle_2_extract.loc[:, 'player'] == 'Sasha Vujacic')]


# df_salary.loc[(df_salary.loc[:, 'player'] == 'Mark Blount')]


# df_salary.loc[(df_salary.loc[:, 'player'] == 'Andrew Bynum')]
# all_kaggle_2_extract.loc[(all_kaggle_2_extract.loc[:, 'player'] == 'Andrew Bynum')]
# (df_salary.loc[:, 'player'] == 'Derrick Rose') & (df_salary.loc[:, 'season'] == 2012)

            player  season      salary
3549  Derrick Rose    2012  17632688.0
4084  Derrick Rose    2013  17632688.0
4464  Derrick Rose    2014  18862876.0
4935  Derrick Rose    2015  20093064.0
5335  Derrick Rose    2016  21323252.0
6231  Derrick Rose    2017    290951.0
1041  Derrick Rose    2018   2176260.0
80    Derrick Rose    2019    $7317074
814   Derrick Rose    2020   7682926.0
98    Derrick Rose    2021   $13445120

            player  season      salary
3549  Derrick Rose    2012  17632688.0


season        player   Age   Tm     G   PER   WS  BPM    TRB    AST  \
22728  2013.0  Derrick Rose  25.0  CHI  10.0   9.7 -0.2 -5.1   32.0   43.0   
23373  2014.0  Derrick Rose  26.0  CHI  51.0  15.9  1.2 -1.3  161.0  251.0   
23973  2015.0  Derrick Rose  27.0  CHI  66.0  13.5  0.4 -3.3  225.0  311.0   
24562  2016.0  Derrick Rose  28.0  NYK  64.0  17.0  3.0 -1.9  246.0  283.0   

        STL   BLK    TOV     PTS  
22728   5.0   1.0   34.0   159.0  
23373  36.0  16.0  161.0   904.0  
23973  44.0  14.0  175.0  1080.0  
24562  44.0  17.0  147.0  1154.0

In [655]:
print (df_temp_salary_kaggle2.loc[df_temp_salary_kaggle2.isnull().any(axis=1)].loc[df_temp_salary_kaggle2.season > 2016])
df_temp_salary_kaggle2.loc[df_temp_salary_kaggle2.isnull().any(axis=1)].loc[df_temp_salary_kaggle2.season > 2016]

                 player  season      salary team  games  age  PER  WS  BPM  \
2737      Stephen Curry    2017  37457154.0  NaN    NaN  NaN  NaN NaN  NaN   
2738         Al Horford    2017  27734406.0  NaN    NaN  NaN  NaN NaN  NaN   
2739    Carmelo Anthony    2017  26243760.0  NaN    NaN  NaN  NaN NaN  NaN   
2740     Damian Lillard    2017  26153057.0  NaN    NaN  NaN  NaN NaN  NaN   
2741    Otto Porter Jr.    2017  26011913.0  NaN    NaN  NaN  NaN NaN  NaN   
...                 ...     ...         ...  ...    ...  ...  ...  ..  ...   
5096         Sam Dekker    2021     $350000  NaN    NaN  NaN  NaN NaN  NaN   
5097      Ish Wainright    2021         NaN  NaN    NaN  NaN  NaN NaN  NaN   
5098      Troy Williams    2021     $122741  NaN    NaN  NaN  NaN NaN  NaN   
5099      Jabari Parker    2021    $2239544  NaN    NaN  NaN  NaN NaN  NaN   
5100  Demetrius Jackson    2021      $92857  NaN    NaN  NaN  NaN NaN  NaN   

      PTS  REB  AST  STL  BLK  TOV  
2737  NaN  NaN  NaN  NaN  

player  season      salary team  games  age  PER  WS  BPM  \
2737      Stephen Curry    2017  37457154.0  NaN    NaN  NaN  NaN NaN  NaN   
2738         Al Horford    2017  27734406.0  NaN    NaN  NaN  NaN NaN  NaN   
2739    Carmelo Anthony    2017  26243760.0  NaN    NaN  NaN  NaN NaN  NaN   
2740     Damian Lillard    2017  26153057.0  NaN    NaN  NaN  NaN NaN  NaN   
2741    Otto Porter Jr.    2017  26011913.0  NaN    NaN  NaN  NaN NaN  NaN   
...                 ...     ...         ...  ...    ...  ...  ...  ..  ...   
5096         Sam Dekker    2021     $350000  NaN    NaN  NaN  NaN NaN  NaN   
5097      Ish Wainright    2021         NaN  NaN    NaN  NaN  NaN NaN  NaN   
5098      Troy Williams    2021     $122741  NaN    NaN  NaN  NaN NaN  NaN   
5099      Jabari Parker    2021    $2239544  NaN    NaN  NaN  NaN NaN  NaN   
5100  Demetrius Jackson    2021      $92857  NaN    NaN  NaN  NaN NaN  NaN   

      PTS  REB  AST  STL  BLK  TOV  
2737  NaN  NaN  NaN  NaN  NaN  NaN  
2738  NaN  NaN  NaN  NaN  NaN  NaN  
2739  NaN  NaN  NaN  NaN  NaN  NaN  
2740  NaN  NaN  NaN  NaN  NaN  NaN  
2741  NaN  NaN  NaN  NaN  NaN  NaN  
...   ...  ...  ...  ...  ...  ...  
5096  NaN  NaN  NaN  NaN  NaN  NaN  
5097  NaN  NaN  NaN  NaN  NaN  NaN  
5098  NaN  NaN  NaN  NaN  NaN  NaN  
5099  NaN  NaN  NaN  NaN  NaN  NaN  
5100  NaN  NaN  NaN  NaN  NaN  NaN  

[2364 rows x 15 columns]

In [656]:
# print (df_salary.loc[(df_salary.loc[:, 'player'] == 'Stephen Curry')])
# print ()
# print (df_salary.loc[(df_salary.loc[:, 'player'] == 'Stephen Curry') & (df_salary.loc[:, 'season'] == 2017)])
# all_kaggle_2_extract.loc[(all_kaggle_2_extract.loc[:, 'player'] == 'Stephen Curry')]

print (df_salary.loc[(df_salary.loc[:, 'player'] == 'Al Horford')])
print ()
print (df_salary.loc[(df_salary.loc[:, 'player'] == 'Al Horford') & (df_salary.loc[:, 'season'] == 2017)])
all_kaggle_2_extract.loc[(all_kaggle_2_extract.loc[:, 'player'] == 'Al Horford')]

          player  season      salary
3586  Al Horford    2012  12000000.0
4117  Al Horford    2013  12000000.0
4503  Al Horford    2014  12000000.0
4949  Al Horford    2015  12000000.0
5352  Al Horford    2016  26540100.0
5793  Al Horford    2017  27734406.0
938   Al Horford    2018  28928710.0
44    Al Horford    2019   $28000000
855   Al Horford    2020  27500000.0
40    Al Horford    2021   $27000000

          player  season      salary
5793  Al Horford    2017  27734406.0


season      player   Age   Tm     G   PER   WS  BPM    TRB    AST  \
21918  2012.0  Al Horford  26.0  ATL  74.0  19.8  8.8  3.1  757.0  240.0   
22534  2013.0  Al Horford  27.0  ATL  29.0  22.0  2.8  2.8  244.0   76.0   
23139  2014.0  Al Horford  28.0  ATL  76.0  21.4  8.7  3.8  544.0  244.0   
23762  2015.0  Al Horford  29.0  ATL  82.0  19.4  9.4  4.1  596.0  263.0   
24343  2016.0  Al Horford  30.0  BOS  68.0  17.7  6.3  3.1  464.0  337.0   

        STL    BLK    TOV     PTS  
21918  78.0   78.0  147.0  1289.0  
22534  27.0   44.0   64.0   538.0  
23139  68.0   98.0  100.0  1156.0  
23762  68.0  121.0  107.0  1249.0  
24343  52.0   87.0  116.0   952.0

### Predictor second investigation 
- df_temp_salary_kaggle2
    - from 2012 ~ 2021
    - 5101 rows
    - 2707 null data in several columns due to missing data (2012 ~ 2016: 343; 2017 ~ : 2364)

- all_kaggle1
    - from 2012 ~ 2020
    - 4544 rows
    - no null data
- combine method
    - 


#### Conclusion
- 



In [657]:
print (df_temp_salary_kaggle2)
print()
# all_kaggle1.isnull().sum()
all_kaggle1

                 player  season      salary team  games   age   PER    WS  \
0           Kobe Bryant    2012  30453805.0  LAL   78.0  34.0  23.0  10.9   
1            Chris Paul    2012  18668431.0  LAC   70.0  27.0  26.4  13.9   
2          Kevin Durant    2012  17832626.0  OKC   81.0  24.0  28.3  18.9   
3          Derrick Rose    2012  17632688.0  NaN    NaN   NaN   NaN   NaN   
4        Deron Williams    2012  17177795.0  BRK   78.0  28.0  20.3  10.9   
...                 ...     ...         ...  ...    ...   ...   ...   ...   
5096         Sam Dekker    2021     $350000  NaN    NaN   NaN   NaN   NaN   
5097      Ish Wainright    2021         NaN  NaN    NaN   NaN   NaN   NaN   
5098      Troy Williams    2021     $122741  NaN    NaN   NaN   NaN   NaN   
5099      Jabari Parker    2021    $2239544  NaN    NaN   NaN   NaN   NaN   
5100  Demetrius Jackson    2021      $92857  NaN    NaN   NaN   NaN   NaN   

      BPM   PTS  REB  AST  STL  BLK  TOV  
0     4.7  27.3  5.6  6.0  1.4  

player team   age  games   PTS   REB  AST  NET_RATING  \
7156           Nolan Smith  POR  24.0     40   2.8   0.7  0.9        -4.2   
7157        Nikola Vucevic  ORL  22.0     77  13.1  11.9  1.9        -8.1   
7158           Norris Cole  MIA  24.0     80   5.6   1.6  2.1         3.1   
7159             Ray Allen  MIA  37.0     79  10.9   2.7  1.7         5.5   
7160             Omer Asik  HOU  26.0     82  10.1  11.7  0.9         5.9   
...                    ...  ...   ...    ...   ...   ...  ...         ...   
11695  Matthew Dellavedova  CLE  30.0     13   2.8   1.8  4.5        -3.1   
11696     Maurice Harkless  SAC  28.0     37   5.2   2.4  1.2        -2.9   
11697            Max Strus  MIA  25.0     39   6.1   1.1  0.6        -4.2   
11698    Marcus Morris Sr.  LAC  31.0     57  13.4   4.1  1.0         4.2   
11699         Aaron Gordon  DEN  25.0     50  12.4   5.7  3.2         2.1   

       season  
7156     2012  
7157     2012  
7158     2012  
7159     2012  
7160     2012  
...       ...  
11695    2020  
11696    2020  
11697    2020  
11698    2020  
11699    2020  

[4544 rows x 9 columns]

In [658]:
# df_temp_salary_kaggle2 = pd.merge(df_salary, all_kaggle_2_extract, on=['player','season'], how = 'left')

df_temp_salary_kaggle2and1 = pd.merge(df_temp_salary_kaggle2, all_kaggle1, on=['player','season', 'team','age', 'games', 'PTS', 'REB', 'AST'], how = 'left')
df_temp_salary_kaggle2and1


player  season      salary team  games   age   PER    WS  \
0           Kobe Bryant    2012  30453805.0  LAL   78.0  34.0  23.0  10.9   
1            Chris Paul    2012  18668431.0  LAC   70.0  27.0  26.4  13.9   
2          Kevin Durant    2012  17832626.0  OKC   81.0  24.0  28.3  18.9   
3          Derrick Rose    2012  17632688.0  NaN    NaN   NaN   NaN   NaN   
4        Deron Williams    2012  17177795.0  BRK   78.0  28.0  20.3  10.9   
...                 ...     ...         ...  ...    ...   ...   ...   ...   
5096         Sam Dekker    2021     $350000  NaN    NaN   NaN   NaN   NaN   
5097      Ish Wainright    2021         NaN  NaN    NaN   NaN   NaN   NaN   
5098      Troy Williams    2021     $122741  NaN    NaN   NaN   NaN   NaN   
5099      Jabari Parker    2021    $2239544  NaN    NaN   NaN   NaN   NaN   
5100  Demetrius Jackson    2021      $92857  NaN    NaN   NaN   NaN   NaN   

      BPM   PTS  REB  AST  STL  BLK  TOV  NET_RATING  
0     4.7  27.3  5.6  6.0  1.4  0.3  3.7         2.7  
1     7.5  16.9  3.7  9.7  2.4  0.1  2.3         NaN  
2     7.7  28.1  7.9  4.6  1.4  1.3  3.5        12.7  
3     NaN   NaN  NaN  NaN  NaN  NaN  NaN         NaN  
4     3.5  18.9  3.0  7.7  1.0  0.4  2.8         NaN  
...   ...   ...  ...  ...  ...  ...  ...         ...  
5096  NaN   NaN  NaN  NaN  NaN  NaN  NaN         NaN  
5097  NaN   NaN  NaN  NaN  NaN  NaN  NaN         NaN  
5098  NaN   NaN  NaN  NaN  NaN  NaN  NaN         NaN  
5099  NaN   NaN  NaN  NaN  NaN  NaN  NaN         NaN  
5100  NaN   NaN  NaN  NaN  NaN  NaN  NaN         NaN  

[5101 rows x 16 columns]

In [659]:
print (df_temp_salary_kaggle2and1.isnull().sum())
print (df_temp_salary_kaggle2and1.loc[(df_temp_salary_kaggle2and1.season <= 2016) & (df_temp_salary_kaggle2and1.NET_RATING.notnull())])
print (df_temp_salary_kaggle2and1.loc[(df_temp_salary_kaggle2and1.season <= 2016) & (df_temp_salary_kaggle2and1.NET_RATING.isnull())])

print(df_temp_salary_kaggle2and1.loc[(df_temp_salary_kaggle2and1.season > 2016) & (df_temp_salary_kaggle2and1.NET_RATING.notnull())])
df_temp_salary_kaggle2and1.loc[(df_temp_salary_kaggle2and1.season > 2016) & (df_temp_salary_kaggle2and1.NET_RATING.isnull())]




player           0
season           0
salary          25
team          2706
games         2706
age           2706
PER           2707
WS            2706
BPM           2706
PTS           2706
REB           2706
AST           2706
STL           2706
BLK           2706
TOV           2706
NET_RATING    4335
dtype: int64
              player  season      salary team  games   age   PER    WS   BPM  \
0        Kobe Bryant    2012  30453805.0  LAL   78.0  34.0  23.0  10.9   4.7   
2       Kevin Durant    2012  17832626.0  OKC   81.0  24.0  28.3  18.9   7.7   
6        Paul Pierce    2012  16790345.0  BOS   77.0  35.0  19.1   7.2   3.4   
7      Zach Randolph    2012  16500000.0  MEM   76.0  31.0  17.9   7.9   0.2   
11     Carlos Boozer    2012  15300000.0  CHI   79.0  31.0  17.1   5.7  -0.6   
...              ...     ...         ...  ...    ...   ...   ...   ...   ...   
2716   Justin Harper    2016     57672.0  PHI    3.0  27.0   4.9  -0.1  -8.8   
2722    Jarrett Jack    2016     57672.0  N

player  season      salary team  games  age  PER  WS  BPM  \
2737      Stephen Curry    2017  37457154.0  NaN    NaN  NaN  NaN NaN  NaN   
2738         Al Horford    2017  27734406.0  NaN    NaN  NaN  NaN NaN  NaN   
2739    Carmelo Anthony    2017  26243760.0  NaN    NaN  NaN  NaN NaN  NaN   
2740     Damian Lillard    2017  26153057.0  NaN    NaN  NaN  NaN NaN  NaN   
2741    Otto Porter Jr.    2017  26011913.0  NaN    NaN  NaN  NaN NaN  NaN   
...                 ...     ...         ...  ...    ...  ...  ...  ..  ...   
5096         Sam Dekker    2021     $350000  NaN    NaN  NaN  NaN NaN  NaN   
5097      Ish Wainright    2021         NaN  NaN    NaN  NaN  NaN NaN  NaN   
5098      Troy Williams    2021     $122741  NaN    NaN  NaN  NaN NaN  NaN   
5099      Jabari Parker    2021    $2239544  NaN    NaN  NaN  NaN NaN  NaN   
5100  Demetrius Jackson    2021      $92857  NaN    NaN  NaN  NaN NaN  NaN   

      PTS  REB  AST  STL  BLK  TOV  NET_RATING  
2737  NaN  NaN  NaN  NaN  NaN  NaN         NaN  
2738  NaN  NaN  NaN  NaN  NaN  NaN         NaN  
2739  NaN  NaN  NaN  NaN  NaN  NaN         NaN  
2740  NaN  NaN  NaN  NaN  NaN  NaN         NaN  
2741  NaN  NaN  NaN  NaN  NaN  NaN         NaN  
...   ...  ...  ...  ...  ...  ...         ...  
5096  NaN  NaN  NaN  NaN  NaN  NaN         NaN  
5097  NaN  NaN  NaN  NaN  NaN  NaN         NaN  
5098  NaN  NaN  NaN  NaN  NaN  NaN         NaN  
5099  NaN  NaN  NaN  NaN  NaN  NaN         NaN  
5100  NaN  NaN  NaN  NaN  NaN  NaN         NaN  

[2364 rows x 16 columns]

### Predictor - 2: try concatenate all predictors dataset, and then merge with the response variable

#### Conclusion:
- all_kaggle1_before16 has more data points (players) compared with all_kaggle_2_extract, and we decided to only collect those players who have full data (inner), instead of getting missed data for the additional data points (players) in all_kaggle1_before16

- shall we use outer, and do missing data impution isntead?


In [715]:
all_kaggle1_before16 = all_kaggle1.loc[all_kaggle1.season <=2016]
all_kaggle1_before16.player.value_counts()
all_kaggle1_before16.loc[all_kaggle1_before16.player == 'AJ Price']



player team   age  games  PTS  REB  AST  NET_RATING  season
7421  AJ Price  WAS  26.0     57  7.7  2.0  3.6        -2.3    2012
7888  AJ Price  MIN  27.0     28  1.6  0.4  0.5        -2.4    2013
8380  AJ Price  CLE  28.0     26  5.1  1.2  1.8       -12.2    2014

In [667]:
all_kaggle1_after16 = all_kaggle1.loc[all_kaggle1.season > 2016]
all_kaggle1_after16


player team   age  games   PTS  REB  AST  NET_RATING  \
9561         Briante Weber  MEM  25.0     18   2.8  1.9  1.2       -21.6   
9562           Jae Crowder  UTA  27.0     80   9.7  3.4  1.2         0.6   
9563         Jahlil Okafor  BKN  22.0     28   6.3  3.0  0.4       -22.6   
9564           Jake Layman  POR  24.0     35   1.0  0.5  0.3       -11.5   
9565          Jakob Poeltl  TOR  22.0     82   6.9  4.8  0.7         8.3   
...                    ...  ...   ...    ...   ...  ...  ...         ...   
11695  Matthew Dellavedova  CLE  30.0     13   2.8  1.8  4.5        -3.1   
11696     Maurice Harkless  SAC  28.0     37   5.2  2.4  1.2        -2.9   
11697            Max Strus  MIA  25.0     39   6.1  1.1  0.6        -4.2   
11698    Marcus Morris Sr.  LAC  31.0     57  13.4  4.1  1.0         4.2   
11699         Aaron Gordon  DEN  25.0     50  12.4  5.7  3.2         2.1   

       season  
9561     2017  
9562     2017  
9563     2017  
9564     2017  
9565     2017  
...       ...  
11695    2020  
11696    2020  
11697    2020  
11698    2020  
11699    2020  

[2139 rows x 9 columns]

In [662]:
all_kaggle_2_extract.rename(columns={'Age': 'age', 'Tm': 'team', 'G': 'games', 'TRB': 'REB'}, inplace=True)
all_kaggle_2_extract = all_kaggle_2_extract[['player', 'season','team', 'games', 'age', 'PER', 'WS', 'BPM','PTS', 'REB', 'AST', 'STL', 'BLK', 'TOV']]
all_kaggle_2_extract['PTS'] = round(all_kaggle_2_extract['PTS'] / all_kaggle_2_extract['games'], 1)
all_kaggle_2_extract['REB'] = round(all_kaggle_2_extract['REB'] / all_kaggle_2_extract['games'], 1)
all_kaggle_2_extract['AST'] = round(all_kaggle_2_extract['AST'] / all_kaggle_2_extract['games'], 1)
all_kaggle_2_extract['STL'] = round(all_kaggle_2_extract['STL'] / all_kaggle_2_extract['games'], 1)
all_kaggle_2_extract['BLK'] = round(all_kaggle_2_extract['BLK'] / all_kaggle_2_extract['games'], 1)
all_kaggle_2_extract['TOV'] = round(all_kaggle_2_extract['TOV'] / all_kaggle_2_extract['games'], 1)
all_kaggle_2_extract

player  season team  games   age   PER   WS  BPM   PTS  REB  \
21679         Quincy Acy  2012.0  TOR   29.0  22.0  15.9  1.1  0.7   4.0  2.7   
21680        Jeff Adrien  2012.0  CHA   52.0  26.0  13.4  1.0 -3.3   4.0  3.8   
21681      Arron Afflalo  2012.0  ORL   64.0  27.0  13.0  2.0 -2.3  16.5  3.7   
21682       Josh Akognon  2012.0  DAL    3.0  26.0  15.3  0.0 -0.2   1.7  0.3   
21683       Cole Aldrich  2012.0  TOT   45.0  24.0  11.1  0.6 -4.0   2.2  2.7   
...                  ...     ...  ...    ...   ...   ...  ...  ...   ...  ...   
24686        Cody Zeller  2016.0  CHO   62.0  24.0  16.7  5.6  2.1  10.3  6.5   
24687       Tyler Zeller  2016.0  BOS   51.0  27.0  13.0  1.0 -2.5   3.5  2.4   
24688  Stephen Zimmerman  2016.0  ORL   19.0  20.0   7.3  0.0 -7.3   1.2  1.8   
24689        Paul Zipser  2016.0  CHI   44.0  22.0   6.9  0.5 -3.7   5.5  2.8   
24690        Ivica Zubac  2016.0  LAL   38.0  19.0  17.0  1.1 -2.5   7.5  4.2   

       AST  STL  BLK  TOV  
21679  0.4  0.4  0.5  0.6  
21680  0.7  0.3  0.5  0.6  
21681  3.2  0.6  0.2  2.2  
21682  0.3  0.0  0.0  0.0  
21683  0.2  0.1  0.5  0.5  
...    ...  ...  ...  ...  
24686  1.6  1.0  0.9  1.0  
24687  0.8  0.1  0.4  0.4  
24688  0.2  0.1  0.3  0.2  
24689  0.8  0.3  0.4  0.9  
24690  0.8  0.4  0.9  0.8  

[3008 rows x 14 columns]

In [714]:
all_kaggle_2_extract.player.value_counts()
all_kaggle_2_extract.loc[all_kaggle_2_extract.player == 'AJ Price']


Empty DataFrame
Columns: [player, season, team, games, age, PER, WS, BPM, PTS, REB, AST, STL, BLK, TOV]
Index: []

- shall not use left join as left and right shall take inner, not outer or others
- shall use either merge or join

In [713]:
df_temp_kaggle_1and2_before2016 = pd.merge(all_kaggle1_before16, all_kaggle_2_extract, on=['player','season', 'team','age', 'games', 'PTS', 'REB', 'AST'], how="inner")
print (len(df_temp_kaggle_1and2_before2016))
df_temp_kaggle_1and2_before2016.loc[df_temp_kaggle_1and2_before2016.player == 'AJ Price']


941


Empty DataFrame
Columns: [player, team, age, games, PTS, REB, AST, NET_RATING, season, PER, WS, BPM, STL, BLK, TOV]
Index: []

In [712]:
test = pd.merge(all_kaggle1_before16, all_kaggle_2_extract, on=['player','season', 'team','age', 'games', 'PTS', 'REB', 'AST'], how="outer")
print (len(test))
test.loc[test.player == 'AJ Price']



4472


player team   age  games  PTS  REB  AST  NET_RATING  season  PER  WS  \
265   AJ Price  WAS  26.0     57  7.7  2.0  3.6        -2.3    2012  NaN NaN   
732   AJ Price  MIN  27.0     28  1.6  0.4  0.5        -2.4    2013  NaN NaN   
1224  AJ Price  CLE  28.0     26  5.1  1.2  1.8       -12.2    2014  NaN NaN   

      BPM  STL  BLK  TOV  
265   NaN  NaN  NaN  NaN  
732   NaN  NaN  NaN  NaN  
1224  NaN  NaN  NaN  NaN

In [716]:
test_941 = test.iloc[:941,:]
print (len(test_941), len(df_temp_kaggle_1and2_before2016))
df_temp_kaggle_1and2_before2016.sort_values('player', inplace = True)
test_941.sort_values('player', inplace = True)
print (df_temp_kaggle_1and2_before2016)
print (test_941)
# diff = (df_temp_kaggle_1and2_before2016 != test_941).any(1)
# df_temp_kaggle_1and2_before2016.loc[diff]


941 941
            player team   age  games   PTS   REB  AST  NET_RATING  season  \
478   Aaron Brooks  CHI  30.0     82  11.6   2.0  3.2         5.2    2014   
638   Aaron Brooks  CHI  31.0     69   7.1   1.5  2.6        -1.4    2015   
884   Aaron Brooks  IND  32.0     65   5.0   1.1  1.9        -3.0    2016   
477   Aaron Gordon  ORL  19.0     47   5.2   3.6  0.7        -5.6    2014   
639   Aaron Gordon  ORL  20.0     78   9.2   6.5  1.6        -1.3    2015   
..             ...  ...   ...    ...   ...   ...  ...         ...     ...   
241  Zach Randolph  MEM  32.0     79  17.4  10.1  2.5         1.5    2013   
479  Zach Randolph  MEM  33.0     71  16.1  10.5  2.2         6.3    2014   
576  Zach Randolph  MEM  34.0     68  15.3   7.8  2.1        -2.0    2015   
882  Zach Randolph  MEM  35.0     73  14.1   8.2  1.7        -1.1    2016   
43   Zach Randolph  MEM  31.0     76  15.4  11.2  1.4         4.6    2012   

      PER   WS  BPM  STL  BLK  TOV  
478  14.4  3.3 -1.4  0.7  0.2 

/Users/tegochang/opt/miniconda3/lib/python3.9/site-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


In [677]:
df_temp_kaggle_1and2 = pd.concat([df_temp_kaggle_1and2_before2016, all_kaggle1_after16], axis=0, join="outer")
df_temp_kaggle_1and2


player team   age  games   PTS   REB  AST  NET_RATING  \
0              Nolan Smith  POR  24.0     40   2.8   0.7  0.9        -4.2   
1           Nikola Vucevic  ORL  22.0     77  13.1  11.9  1.9        -8.1   
2                Ray Allen  MIA  37.0     79  10.9   2.7  1.7         5.5   
3                Omer Asik  HOU  26.0     82  10.1  11.7  0.9         5.9   
4          Rasheed Wallace  NYK  38.0     21   7.0   4.0  0.3         3.7   
...                    ...  ...   ...    ...   ...   ...  ...         ...   
11695  Matthew Dellavedova  CLE  30.0     13   2.8   1.8  4.5        -3.1   
11696     Maurice Harkless  SAC  28.0     37   5.2   2.4  1.2        -2.9   
11697            Max Strus  MIA  25.0     39   6.1   1.1  0.6        -4.2   
11698    Marcus Morris Sr.  LAC  31.0     57  13.4   4.1  1.0         4.2   
11699         Aaron Gordon  DEN  25.0     50  12.4   5.7  3.2         2.1   

       season   PER   WS  BPM  STL  BLK  TOV  
0        2012   7.1 -0.5 -8.6  0.2  0.0  0.7  
1        2012  17.8  5.7  0.4  0.8  1.0  1.8  
2        2012  14.7  5.4  0.2  0.8  0.2  1.3  
3        2012  14.9  5.5 -1.3  0.6  1.1  2.1  
4        2012  16.7  0.6 -3.2  0.6  0.7  0.5  
...       ...   ...  ...  ...  ...  ...  ...  
11695    2020   NaN  NaN  NaN  NaN  NaN  NaN  
11696    2020   NaN  NaN  NaN  NaN  NaN  NaN  
11697    2020   NaN  NaN  NaN  NaN  NaN  NaN  
11698    2020   NaN  NaN  NaN  NaN  NaN  NaN  
11699    2020   NaN  NaN  NaN  NaN  NaN  NaN  

[3080 rows x 15 columns]

### More data for predictors

- Kaggle_3: PlayerStatisticsPerGame.csv
    - season: 1950 ~ 2019/20 
- Kaggle_4: nba_full_final.csv
    - season: 1950 ~ 2019/20 
    - Advanced version of Kaggle_3

In [723]:
df_kaggle_3 = pd.read_csv("../dataset/age/archive-2/PlayerStatisticsPerGame.csv")
print (df_kaggle_3.columns)
df_kaggle_3.loc[df_kaggle_3.Player == "Tyler Zeller"].sort_values('year')

Index(['Player', 'Pos', 'Age', 'Tm', 'G', 'GS', 'MP', 'FG', 'FGA', 'FG%', '3P',
       '3PA', '3P%', '2P', '2PA', '2P%', 'eFG%', 'FT', 'FTA', 'FT%', 'ORB',
       'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS', 'year'],
      dtype='object')


Player Pos  Age   Tm     G    GS    MP   FG  FGA    FG%  ...  \
25644  Tyler Zeller   C   23  CLE  77.0  55.0  26.4  3.2  7.3  0.438  ...   
18464  Tyler Zeller   C   24  CLE  70.0   9.0  15.0  2.2  4.1  0.538  ...   
19011  Tyler Zeller   C   25  BOS  82.0  59.0  21.1  4.1  7.5  0.549  ...   
19510  Tyler Zeller   C   26  BOS  60.0   3.0  11.8  2.3  4.8  0.476  ...   
20005  Tyler Zeller   C   27  BOS  51.0   5.0  10.3  1.5  3.1  0.494  ...   
20539  Tyler Zeller   C   28  TOT  66.0  34.0  16.8  2.8  5.1  0.560  ...   
20540  Tyler Zeller   C   28  BRK  42.0  33.0  16.7  3.0  5.5  0.546  ...   
20541  Tyler Zeller   C   28  MIL  24.0   1.0  16.9  2.6  4.4  0.590  ...   
21129  Tyler Zeller   C   29  TOT   6.0   1.0  15.5  2.7  5.0  0.533  ...   
21130  Tyler Zeller   C   29  ATL   2.0   0.0   5.5  0.0  1.0  0.000  ...   
21131  Tyler Zeller   C   29  MEM   4.0   1.0  20.5  4.0  7.0  0.571  ...   

       ORB  DRB  TRB  AST  STL  BLK  TOV   PF   PTS  year  
25644  2.1  3.6  5.7  1.2  0.5  0.9  1.2  3.3   7.9  2013  
18464  1.5  2.6  4.0  0.5  0.3  0.5  0.9  2.0   5.7  2014  
19011  1.8  3.9  5.7  1.4  0.2  0.6  0.9  2.5  10.2  2015  
19510  1.0  1.9  3.0  0.5  0.2  0.4  0.8  1.6   6.1  2016  
20005  0.8  1.6  2.4  0.8  0.1  0.4  0.4  1.2   3.5  2017  
20539  1.7  3.0  4.6  0.7  0.2  0.5  0.7  1.9   6.7  2018  
20540  1.5  3.1  4.6  0.7  0.2  0.5  0.8  1.9   7.1  2018  
20541  2.0  2.7  4.6  0.8  0.3  0.6  0.5  2.0   5.9  2018  
21129  1.8  2.2  4.0  0.7  0.2  0.5  0.7  3.3   7.7  2019  
21130  1.0  2.0  3.0  0.5  0.0  0.0  0.0  2.0   0.0  2019  
21131  2.3  2.3  4.5  0.8  0.3  0.8  1.0  4.0  11.5  2019  

[11 rows x 30 columns]

In [726]:
df_kaggle_4 = pd.read_csv("../dataset/age/nba_full_final.csv")
print (df_kaggle_4.columns)
df_kaggle_4.loc[df_kaggle_4.player == "Tyler Zeller"].sort_values('year')

Index(['Unnamed: 0', 'player', 'pos', 'age', 'tm', 'g', 'gs', 'mp', 'fg',
       'fga', 'fg%', '3p', '3pa', '3p%', '2p', '2pa', '2p%', 'efg%', 'ft',
       'fta', 'ft%', 'orb', 'drb', 'trb', 'ast', 'stl', 'blk', 'tov', 'pf',
       'pts', 'year'],
      dtype='object')


Unnamed: 0        player pos  age   tm   g    gs    mp   fg  fga  ...  \
18330       18376  Tyler Zeller   C   23  CLE  77  55.0  26.4  3.2  7.3  ...   
18810       18858  Tyler Zeller   C   24  CLE  70   9.0  15.0  2.2  4.1  ...   
19300       19350  Tyler Zeller   C   25  BOS  82  59.0  21.1  4.1  7.5  ...   
19775       19826  Tyler Zeller   C   26  BOS  60   3.0  11.8  2.3  4.8  ...   
20257       20309  Tyler Zeller   C   27  BOS  51   5.0  10.3  1.5  3.1  ...   
20794       20849  Tyler Zeller   C   28  TOT  66  34.0  16.8  2.8  5.1  ...   
21321       21380  Tyler Zeller   C   29  TOT   6   1.0  15.5  2.7  5.0  ...   
21848       21909  Tyler Zeller   C   30  SAS   2   NaN   2.0  0.5  2.0  ...   

       orb  drb  trb  ast  stl  blk  tov   pf   pts     year  
18330  2.1  3.6  5.7  1.2  0.5  0.9  1.2  3.3   7.9  2012-13  
18810  1.5  2.6  4.0  0.5  0.3  0.5  0.9  2.0   5.7  2013-14  
19300  1.8  3.9  5.7  1.4  0.2  0.6  0.9  2.5  10.2  2014-15  
19775  1.0  1.9  3.0  0.5  0.2  0.4  0.8  1.6   6.1  2015-16  
20257  0.8  1.6  2.4  0.8  0.1  0.4  0.4  1.2   3.5  2016-17  
20794  1.7  3.0  4.6  0.7  0.2  0.5  0.7  1.9   6.7  2017-18  
21321  1.8  2.2  4.0  0.7  0.2  0.5  0.7  3.3   7.7  2018-19  
21848  1.5  0.5  2.0  NaN  NaN  NaN  NaN  NaN   1.0  2019-20  

[8 rows x 31 columns]

In [730]:
df_kaggle_5 = pd.read_csv("../dataset/age/NBA Stats 2018-2021/NBA_stats.csv")
print (df_kaggle_5.columns)
df_kaggle_5.loc[df_kaggle_5.Player == "Tyler Zeller"].sort_values('Year')

Index(['Player', 'Pos', 'Age', 'Team', 'Games', 'Minutes Played',
       'Fields Goal', 'Fields Goal Attempted', '3-points Field Goal',
       '3-points Field Goal Attempted', '2-points Field Goal',
       '2-points Field Goal Attempted', 'Free Throws', 'Free Throws Attempted',
       'Offensive Rebounds', 'Defensive Rebounds', 'Total Rebounds', 'Assists',
       'Steals', 'Blocks', 'Turnovers', 'Personal Fouls', 'Points', 'Rank',
       'Year'],
      dtype='object')


Player Pos  Age Team  Games  Minutes Played  Fields Goal  \
658   Tyler Zeller   C   28  TOT     66            16.8          2.8   
659   Tyler Zeller   C   28  BRK     42            16.7          3.0   
660   Tyler Zeller   C   28  MIL     24            16.9          2.6   
1365  Tyler Zeller   C   29  TOT      6            15.5          2.7   
1366  Tyler Zeller   C   29  ATL      2             5.5          0.0   
1367  Tyler Zeller   C   29  MEM      4            20.5          4.0   
2020  Tyler Zeller   C   30  SAS      2             2.0          0.5   

      Fields Goal Attempted  3-points Field Goal  \
658                     5.1                  0.2   
659                     5.5                  0.2   
660                     4.4                  0.0   
1365                    5.0                  0.0   
1366                    1.0                  0.0   
1367                    7.0                  0.0   
2020                    2.0                  0.0   

      3-points Field Goal Attempted  ...  Defensive Rebounds  Total Rebounds  \
658                             0.4  ...                 3.0             4.6   
659                             0.6  ...                 3.1             4.6   
660                             0.1  ...                 2.7             4.6   
1365                            0.2  ...                 2.2             4.0   
1366                            0.5  ...                 2.0             3.0   
1367                            0.0  ...                 2.3             4.5   
2020                            0.0  ...                 0.5             2.0   

      Assists  Steals  Blocks  Turnovers  Personal Fouls  Points  Rank  Year  
658       0.7     0.2     0.5        0.7             1.9     6.7   537  2018  
659       0.7     0.2     0.5        0.8             1.9     7.1   537  2018  
660       0.8     0.3     0.6        0.5             2.0     5.9   537  2018  
1365      0.7     0.2     0.5        0.7             3.3     7.7   528  2019  
1366      0.5     0.0     0.0        0.0             2.0     0.0   528  2019  
1367      0.8     0.3     0.8        1.0             4.0    11.5   528  2019  
2020      0.0     0.0     0.0        0.0             0.0     1.0   527  2020  

[7 rows x 25 columns]

In [737]:
df_nbastat_t_2021 = pd.read_csv("../dataset/player_stat/traditional/2021.csv")
df_nbastat_t_2021 = df_nbastat_t_2021.iloc[:457,:]
df_nbastat_t_2021

Unnamed: 0                 PLAYER TEAM   AGE    GP     W    L   MIN  \
0           1.0           Kevin Durant  BKN  33.0  15.0  10.0  5.0  34.5   
1           2.0          Stephen Curry  GSW  33.0  14.0  12.0  2.0  33.5   
2           3.0            Paul George  LAC  31.0  14.0   9.0  5.0  35.0   
3           4.0          DeMar DeRozan  CHI  32.0  14.0  10.0  4.0  35.3   
4           5.0  Giannis Antetokounmpo  MIL  26.0  13.0   6.0  7.0  32.9   
..          ...                    ...  ...   ...   ...   ...  ...   ...   
452       431.0             Sam Dekker  TOR  27.0   1.0   1.0  0.0   0.9   
453       431.0            Serge Ibaka  LAC  32.0   2.0   2.0  0.0   6.9   
454       431.0       Vernon Carey Jr.  CHA  20.0   1.0   1.0  0.0   0.8   
455       431.0           Wayne Selden  NYK  27.0   1.0   1.0  0.0   1.0   
456       431.0              Yves Pons  MEM  22.0   1.0   0.0  1.0   2.8   

      PTS   FGM  ...   REB  AST  TOV  STL  BLK   PF    FP  DD2  TD3   +/-  
0    28.9  10.7  ...   8.1  5.0  3.1  0.5  0.5  1.2  46.3  6.0  1.0   3.7  
1    28.7   9.2  ...   6.3  6.6  3.2  1.7  0.7  1.7  50.2  5.0  1.0  13.9  
2    27.0   9.8  ...   8.2  5.1  4.4  2.2  0.4  2.6  48.0  5.0  0.0   4.0  
3    26.9   9.6  ...   5.4  4.1  1.9  0.9  0.4  2.3  41.6  1.0  0.0   8.5  
4    26.5   9.5  ...  11.3  6.0  3.0  1.1  1.7  3.2  54.4  8.0  0.0   5.0  
..    ...   ...  ...   ...  ...  ...  ...  ...  ...   ...  ...  ...   ...  
452   0.0   0.0  ...   0.0  0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  -5.0  
453   0.0   0.0  ...   2.0  0.5  0.5  0.0  0.5  2.5   4.2  0.0  0.0   3.5  
454   0.0   0.0  ...   1.0  0.0  0.0  0.0  0.0  0.0   1.2  0.0  0.0   0.0  
455   0.0   0.0  ...   0.0  0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  -2.0  
456   0.0   0.0  ...   1.0  0.0  0.0  0.0  0.0  0.0   1.2  0.0  0.0  -3.0  

[457 rows x 30 columns]

In [739]:
df_salary_2021 = pd.read_csv("../dataset/salary/s2021.csv")
df_salary_2021

2021/22          Unnamed: 1       2021/22       2022/23       2023/24  \
0         1.0      Stephen Curry  $45,780,966   $48,070,014   $51,915,615    
1         2.0       James Harden  $44,310,840   $47,366,760            $0    
2         2.0          John Wall  $44,310,840   $47,366,760            $0    
3         4.0  Russell Westbrook  $44,211,146   $47,063,478            $0    
4         5.0       LeBron James  $41,180,544   $44,474,988            $0    
..        ...                ...           ...           ...           ...   
532     533.0      Troy Williams     $122,741            $0            $0    
533     534.0     Mamadi Diakite     $100,000            $0            $0    
534     535.0  Demetrius Jackson      $92,857       $92,857       $92,857    
535     536.0  Freddie Gillespie      $50,000            $0            $0    
536       NaN                NaN           NaN           NaN           NaN   

          2024/25  
0    $55,761,217   
1             $0   
2             $0   
3             $0   
4             $0   
..            ...  
532           $0   
533           $0   
534           $0   
535           $0   
536           NaN  

[537 rows x 6 columns]

In [ ]:
trim_contract_bb.rename(columns={'Player': 'player', 'season': 'season', '2021-22': 'salary'}, inplace=True)


trim_contract_bb = trim_contract_bb[['Player', 'season', '2021-22']]
trim_contract_bb.loc[:,'player'] = trim_contract_bb.loc[:,'player'].str.split('\\', expand=True)[0]
trim_contract_sql.loc[:,'season'] = trim_contract_sql.loc[:,'season'].str.split('-', expand=True)[0]
trim_contract_bb['season'] = pd.to_numeric(trim_contract_bb['season'])


